### ISSUE TITLE
Change tuning using configuration interface

### ISSUE THREAD

**Issue by gmeik** [Alexwijn/SAT#12]
The default tuning is too aggressive for my underfloor heating system. The gain is so high that the floor heating system saturates (thermostatic valve closes) on a setpoint change of +0.5 °C. This is not necessary or useful because the overall response of the heating system is _very_ slow.
It would be a nice enhancement if it was possible to set the tuning to manual values from the configuration interface or using a service (without the need to re-create the SAT to enter different tuning values). I should have some idea what I am doing, being an advanced process control engineer. ;)
Similarly for the minimum setpoint. It was estimated a bit too low, and the only way I could change this was by re-creating the thermostat. But maybe I missed something...

Could you also explain the control algorithm in more detail in the documentation? I mean the structure of the PID controller, what are its inputs and outputs and how the 'Heating curve coefficient' works.

**Comment by sergeantd83** [comment]
Hi @gmeik thanks for trying SAT! SAT CONTROL algorithm is a transfer function. We use a Laplace transform to calculate the heat losses of the home, as input values we use the outside temperature and the room setpoint. Then we use a series PID algorithm with input values room temperature and room setpoint in order to calculate the error. We use the integral part and the derivative adhoc. That means we enable the integral part only in deadband to anticipate the heating curve value ( Laplace transform ) and the derivative part is used only outside the deadband. Moreover we calculated the fixed Ti and Td values based on the transfer function of the in series PID controller. 
One question: Did you set your heating system during initial configuration as underfloor?

**Comment by gmeik** [comment]
> One question: Did you set your heating system during initial configuration as underfloor?

Yes, it is set it to underfloor. Let me tell the detailed story. I want to use less heating power at night and more power at day, because this increases the level of comfort (floor temperature at day). Otherwise the floor is only barely warm. And I have no plans to increase heat loss to increase floor temperature. :) So I'm changing the setpoint up and down by 0.5 °C. My previous (DIY) PI controller directly PWM'ed the boiler, so it was basically a power / fired duty controller. The boiler is limited to its lowest fired duty setting, 6 kW, which is still excessive, because the house needs at most ~4 kW. The overall heat transfer coefficient of the house was estimated from historical data, so I added a feedforward from ambient temperature to the output with a gain of -3%/°C. The standard/mixed PI controller had a fixed gain of 8% per °C and integral  time of 28800 seconds. This controller heated up the floor slowly in the morning, but it did not saturate of the floor heating system and temperature control was pretty stable, about 0.5 °C deviation from setpoint does not lead to discomfort combined with floor heating. It also didn´t stop the heating completely on a sunny day, preventing cold floor.

And thanks a lot for the explanation! I understand what you write, including the heating curve. What I don´t understand yet is
1. How does the output of the PID controller connect to the heating curve? Are they added together?
2. What is the reasoning behind the variable tuning calculation of the PID controller? I see in the Python code the kp/ki/kd are calculated from a coefficient, but what does it mean? The end result seems a too high gain, so the floor heating system saturates for only +0.5 °C setpoint change...

**Comment by sergeantd83** [comment]
> 1. How does the output of the PID controller connect to the heating curve? Are they added together?

The Contor Setpoint formula ( transfer function ) actually is:
`CS = HCV * ( 1 + PID OUTPUT )`
HCV = Heating curve value ( Laplace transform )
( 1 + PID OUTPUT ) = In series PID controller with K=1 )

As a result the PID controller acts as a factor in the formula.

> What is the reasoning behind the variable tuning calculation of the PID controller?

When you have higher outside temperature the calculated CS is lower and vice versa because everything is related to OT.

> The end result seems a too high gain, so the floor heating system saturates for only +0.5 °C setpoint change...

Can you explain further please?

**Comment by sergeantd83** [comment]
Can you please provide some informations?

1. kP of your previous ( DIY ) controller?
2. Your current heating curve coefficient you use?
3. Your Overshoot protection value ( it is stored as an attribute in the SAT thermostat ), have you ever calculated the min output of your boiler at 0%? If yes is it the same as the calculated overshoot protection from SAT?

**Comment by gmeik** [comment]
The kP of the previous controller was 8% per °C. That is: 8% power output.
The coefficient is set to 1.0.
The overshoot protection value is 37 °C. That is where the system settles at 0% output (6 kW fired duty).

**Comment by sergeantd83** [comment]
The proportional gain for automatic gains is `kP = 1.65 * HCV`. Can you modify the value 1.65 and inform us which one is the best for your system?

**Comment by gmeik** [comment]
I made an extreme change to the proportional gain calculation: `kP = 0.10 * HCV` (since 14 November).
At the same time, I adjusted the heating curve coefficient to 1.8, this gives a good match between supply temperature and outside temperature.
![Thermostat](https://github.com/Alexwijn/SAT/assets/123992613/e832efe4-245d-46ff-9b9d-88dd613d2094)
As you can see, the temperature matches the setpoint quite well for an outside temperature range of 4 - 14 °C. The result is also comfortable to the users (we are used to 18 °C).

The response of the system itself is very slow. That is the nature of heating loops embedded in concrete. If I change the setpoint to 20 °C, I guess it would take a day to get there.

Maybe `kP = 0.10 * HCV` cannot be applied universally to floor heating systems. But I am convinced `kP = 1.65 * HCV` is excessive for low temperature floor heating systems. Imagine `HCV = 30`, then `kP = 1.65 * 30 = 49.5` and a 0.5 °C error would result in a control output of `30 + 25 = 55 °C` (ignoring derivative / integral parts). That is higher than any reasonable outside temperature would require, and completely saturates the system for a time...

Why does `kP` depend on `HCV`, and not on the heating curve coefficient? Controller tuning should be system-dependent, not temperature-dependent... Can we discuss this?

Suggestion:
`kP = x * Coefficient`
where `x` is an 'aggressiveness' factor, which could be user-changed between 1 (very gentle, usable for slow floor heating sytems) and 20 (aggressive, more usable for fast radiator systems) or so. The `kD` needs to be adjusted as well, I need to look into that.

- For my setup with `Coefficient = 1.8`, `x = 1.4` results in `kP = 1.4 * 1.8 = 2.5`
- Another user with radiator setup uses `kP = 35`, he would need a much higher `x`, something like 10

We can solve this using manual tuning of course. But it is called a smart auto-tune thermostat, so let's make it work for floor heating as well!

**Comment by sergeantd83** [comment]
@gmeik thanks for the recommendations. We already added in the the latest commit the “aggressiveness factor”. If you want you can test it.

**Comment by Alexwijn** [comment]
Did this resolve the issue?

**Comment by gmeik** [comment]
I'm still trying it out, but it seems to solve the issue indeed. As a starting point I use an aggressiveness factor of 1.
